src/sqlfluff/core/default_config.cfg

```diff
@@ -143,3 +143,8 @@ unquoted_identifiers_policy = all
 quoted_identifiers_policy = all
 allow_space_in_identifier = False
 additional_allowed_characters = ""
+
+[sqlfluff:rules:L059]
+# Policy on quoted and unquoted identifiers
+force_quote_identifier = False
+preferred_quote_identifier = '"'
```

Comment:
I wanted to dissociate the 2 configurations because having a preferred quoting system doesn't mean you want every identifier to be quoted. This is actually the opposite looking at best practices: the better is to not quote identifiers unless you don't have the choice (special chars in fields, reserved keywords, ...).

Even if many dialects use backtick by default, the SQL-92 ANSI standard is using double quotes (and actually, even MySQL can switch to this default standard with `SET GLOBAL SQL_MODE=ANSI_QUOTES`.

Also, looking at the unit tests already defined in L059 before my changes, the double-quote was already the expected behavior, that's why I wanted to set it by default: doing it this way will break nothing from original behavior.

What I could try, if you find this too dangerous, is adding the capability of adding "default" as a value of `preferred_quote_identifier`. When this is set to "default" value, in the L059 process, I can define that `quote` is evaluated to the default quote usage depending on the dialect.
The only thing I'd miss for this is the knowledge about "what is the default quote identifier character for each dialect?". 
And doing this might make it harder to add new dialects, so I guess someone more focused on this repo than me should approve this "default" idea, first...

Reply:
I think default should be unset (or empty string). That way it's an explicit decision to force one quote type and default is not to do that.

And I'd rename the options, as mentioned below, to the following:

- `force_quote_identifiers` (default: `False`) - replaces unquoted identifiers with quoted version (e..g. foo to "foo")
- `force_quote_type` (default: unset) - Changes quotes to preferred type (e.g. 'foo' to "foo")

Reply:
Checking out BigQuery it's definitely more complicated. The two types of quotes are very different and cannot be interchanged:

```sql
--Valid - the literal a can be in double quotes, and alias can be in backticks
SELECT "a" AS `b`;

--Invalid - the `a` suggests an identifier from the FROM clause.
SELECT `a` AS `b`;

--Invalid - alias cannot be in quotes
SELECT "a" AS "b";

--Valid:
SELECT
  date aS `col1`, -- date column
  "date" aS `col2`, -- literal
  `date` aS `col3` -- date column (same as col1)
FROM `owner.schema.table` AS `tab1`;
```

So consolidating those to only backticks or only double quotes will break the SQL as they mean different things.

Default definitely has to be not set.

Reply:
Ok, I understand what you mean. I'll rework this using your proposal for the option names, and try to not mix both configuration expectations.

Reply:
I don't understand this:
> `force_identifier_quote_type` (default: unset) - Changes quotes to preferred type (e.g. 'foo' to "foo")

If this config is blank (`""`), then it should either:
1. change quotes to blank (e.g. `'foo'` to `foo`). 
2. change quotes to the default one that is dialect dependent (e.g. `'foo'` to ``` `foo` ``` for MySQL, BigQuery, ..., but `'foo'` to `"foo"` for Ansi, ...).

If you wanna go for 1., then to me there's a conflict with configuration `force_quote_identifier = True`, as the identifier is expected to be quoted, but with no characters, so it doesn't make sense.
If you wanna go for 2., I think it will be definitively less dangerous, but I need to know which default quote is to use for every dialects (I might lack of knowledge about it).

Reply:
I'd suggesting if it's unset we don't change quotes and leave them as is (i.e. current L059 functionality).

Reply:
> I'd suggesting if it's unset we don't change quotes and leave them as is (i.e. current L059 functionality).

Will try out this. Sounds good to me.

Reply:
@tunetheweb 
I have some issues in managing the `force_identifier_quote_type` when it's not empty (overwritten by user definition).

If I change your previous example to:
```sql
SELECT
  foo aS `col1`, -- date column
  "foo" aS `col2`, -- literal
  `foo` aS "col3" -- date column (same as col1)
```
if, `"foo"` is parsed as a segment of type `quoted_identifier`, it will be impossible to infer if this is a literal string or a quoted identifier `foo` with `"` as quote, just like it is for `"col3"` (I know this breaks for BigQuery, but it's allowed in lots of other dialects as default quote for SQL-92 ANSI).

I guess I'll just reduce the scope of this enhancement and simply deal with `force_quote_identifier = [True,False]`.
Default to `False` (current behavior on `main` branch). When `True`, the L059 rule will fail if there's some naked_identifiers.

About my check that should ensure which character is used to quote the identifier, that's sound currently impossible to me as long as we can't differentiate literal strings from quoted identifiers with `"`. I'll find another way using business methods that shouldn't be in the scope of SQLFluff tool (business oriented).

Reply:
_Resolved as wanted behavior has changed_